libtransmission/verify.cc

```diff
@@ -50,10 +89,18 @@ void tr_verify_worker::verify_torrent(Mediator& verify_mediator, std::atomic<boo
         auto const file_length = metainfo.file_size(file_index);
 
         /* if we're starting a new file... */
-        if (file_pos == 0 && fd == TR_BAD_SYS_FILE && file_index != prev_file_index)
+        if (file_pos == 0U && fd == TR_BAD_SYS_FILE && file_index != prev_file_index)
         {
-            auto const found = verify_mediator.find_file(file_index);
-            fd = !found ? TR_BAD_SYS_FILE : tr_sys_file_open(found->c_str(), TR_SYS_FILE_READ | TR_SYS_FILE_SEQUENTIAL, 0);
+            if (auto const found = verify_mediator.find_file(file_index); found)
+            {
+                fd = tr_sys_file_open(found->c_str(), TR_SYS_FILE_READ | TR_SYS_FILE_SEQUENTIAL, 0);
+            }
+            else if (file_length == 0U && verify_mediator.file_piece_is_wanted(file_index))
+            {
+                fd = create_empty_file(
```

Comment:
Verifying torrent is strictly read-only operation. No torrent data may be touched, only torrent metadata should be altered. (Torrent might be in status "stopped" and/or torrent data might be made "read-only" before since no changes of data after complete downloading are expected anymore.)

But how can a torrent with 100% downloaded data missing 0 byte file(s) be marked as incomplete? We need to redefine the "completeness". It should be an "and" operation with something like "wanted_files_count". For old torrents wanted_files_count will include 0 bytes files and thus not match "have_files_count", marking torrent as "incomplete".

Then the user if needed could start such torrent (placing it in status "downloading") and even offline transmission can complete it just by creating those missing 0 bytes files.

Reply:
> But how can a torrent with 100% downloaded data missing 0 byte file(s) be marked as incomplete? We need to redefine the "completeness". It should be an "and" operation with something like "wanted_files_count". For old torrents wanted_files_count will include 0 bytes files and thus not match "have_files_count", marking torrent as "incomplete".

Please convince me why I should do it this way.

> Verifying torrent is strictly read-only operation. No torrent data may be touched, only torrent metadata should be altered.

Is this just an idealistic opinion? Or could you name some real benefits of your proposal, and/or drawbacks of the current behaviour?

Practically speaking, your proposal takes more work to implement and debug, yet I don't see any benefits for doing it one way over the other.

FYI, libtorrent-rasterbar also creates empty files when verifying torrents.

Reply:
I am probably missing some context here, but _if_ this means that previously-added torrents that were marked as 'completed' will be changed to 'incomplete' when users upgrade to 4.1.0, that will cause a lot of confusion + bug reports.

I don't have any strong opinions on creation during verification, just a general aversion to confusion & bug reports :smile_cat:

Reply:
I've assumed people missed those 0 byte files on disk in old torrents. If not really, the default can be the opposite: count those files as "have" and only after manual verification mark them as missing.

Unless transmission would rename "verification" into "verification with auto-fix" I'd be surprised if after verification some torrent would change status to failed (cannot create file), since I've just asked to verify it, not to try change some data on disk and I would report this as bug.